### PR TITLE
`Programming exercises`: Improve repository view and allow download

### PR DESCRIPTION
**Pull request by pzdr7**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

The "Open repository" view had a few unnecessary UI elements. It was also not possible to download the repository from there.

### Description

- Removes the top (create files/folders) and bottom (editor/commit state) bar of the file browser in the repository view
- Adds a download button to the repository view
- Adds an endpoint that allows exporting a single student repository (must be owner or at least TA)

### Steps for Testing

Prerequisites:
- Test Server with LocalVC enabled
- 1 Instructor
- 1 Student
- 1 Programming Exercise with at least one participation

**Instructor**

1. Log in as the instructor
2. Navigate to Course Management > Your course > Exercises
3. Click the title of your programming exercise.
4. For the template / solution / test repository, click "Open Repository"
5. In the repository view, verify that you can see the "Download Repository" button and that it works correctly.
6. Go back to the programming exercise and open the "Participations" view.
7. Click "Open repository" on one participation and download the repository.
8. Verify that this works correctly (e.g. in the downloaded repository, view the commit history using `git log`).

**Student**

9. Log in as the student
10. Navigate to Courses > Your course > Your programming exercise
11. Hit "Open repository" and download your repository there. 
12. As before, verify that this works

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [programming-exercise.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6528/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/services/programming-exercise.service.ts.html) | 90.38% | ✅ |
| [programming-exercise-instructor-repo-download.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6528/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/actions/programming-exercise-instructor-repo-download.component.ts.html) | 100% | ✅ |
| [programming-exercise-student-repo-download.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6528/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/actions/programming-exercise-student-repo-download.component.ts.html) | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [GitService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6528/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors/GitService.html) | 85% | ✅ |
| [ProgrammingExerciseExportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6528/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.export/ProgrammingExerciseExportService.html) | 73% | ✅ |
| [ProgrammingExerciseExportImportResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6528/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.programming/ProgrammingExerciseExportImportResource.html) | 91% | ✅ |

### Screenshots

image

image

### CODE REVIEW

**Review by coolchock — DISMISSED** [review]
tested on ts2, works as described. Repositories are downloaded correctly, git inside of them works. It's not possible to download repository, that doesn't belong to the particular student 👍

File: src/main/java/de/tum/in/www1/artemis/web/rest/programming/ProgrammingExerciseExportImportResource.java [hunk 1]
```diff
@@ -529,4 +529,32 @@ public ResponseEntity<Resource> exportStudentRequestedRepository(@PathVariable l
...
+     * GET /programming-exercises/:exerciseId/export-student-repository/:participationId : Exports the repository belonging to a participation as a zip file.
+     *
+     * @param exerciseId      The id of the programming exercise
+     * @param participationId The id of the student participation for which to export the repository.
+     * @return A ResponseEntity containing the zipped repository.
+     * @throws IOException If the repository could not be zipped.
+     */
+    @GetMapping("programming-exercises/{exerciseId}/export-student-repository/{participationId}")
+    @EnforceAtLeastStudent
+    @FeatureToggle({ Feature.ProgrammingExercises, Feature.Exports })
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/programming/ProgrammingExerciseExportImportResource.java:543]: Why do we add this annotation to every method instead of the class here?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Code looks good overall. Left one comment and an optional comment/question:

File: src/main/java/de/tum/in/www1/artemis/web/rest/programming/ProgrammingExerciseExportImportResource.java (hunk 1 above)
Comment (pzdr7) [reply src/main/java/de/tum/in/www1/artemis/web/rest/programming/ProgrammingExerciseExportImportResource.java:543]: Interestingly, I've only ever seen this as a method-level annotation. I wonder what the reason for it is. I'd say it's worth exploring in a separate PR as it seems to affect multiple Resource classes

File: src/main/java/de/tum/in/www1/artemis/service/export/ProgrammingExerciseExportService.java [hunk 2]
```diff
@@ -439,6 +439,39 @@ public Optional<File> exportStudentRequestedRepository(long exerciseId, boolean
...
+     * @param participation The participation for which to export the repository.
+     * @param exportErrors  A list in which to store errors that occur during the export.
+     * @return The zipped repository if the export was successful, otherwise an empty optional.
+     */
+    public Optional<File> exportStudentRepository(long exerciseId, ProgrammingExerciseStudentParticipation participation, List<String> exportErrors) {
+        var exerciseOrEmpty = loadExerciseForRepoExport(exerciseId, exportErrors);
+        if (exerciseOrEmpty.isEmpty()) {
+            var error = "Failed to export student repository because the exercise " + exerciseId + " does not exist.";
+            log.info(error);
+            exportErrors.add(error);
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/java/de/tum/in/www1/artemis/service/export/ProgrammingExerciseExportService.java]: I removed it

**Review by JohannesStoehr — APPROVED** [review]

File: src/main/webapp/app/localvc/repository-view/repository-view.component.ts [hunk 3]
```diff
@@ -79,11 +81,11 @@ export class RepositoryViewComponent implements OnInit, OnDestroy {
             this.participationCouldNotBeFetched = false;
             const exerciseId = Number(params['exerciseId']);
             const participationId = Number(params['participationId']);
-            if (participationId) {
+            this.repositoryType = participationId ? 'USER' : params['repositoryType'];
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/localvc/repository-view/repository-view.component.ts:84]: The way the routes are set up is that we will have:

For the special repositories, e.g.
```
https://artemis-test2.artemis.cit.tum.de/course-management/18/programming-exercises/31/repository/TEMPLATE
```

In which case `params['repositoryType']` will be `'TEMPLATE'`.

For student repositories:

```
https://artemis-test2.artemis.cit.tum.de/course-management/18/programming-exercises/31/participations/97/repository
```

In which case `participationId` will be 97 (i.e., a number >= 1), so params['repositoryType'] won't be accessed (it would be undefined here).

TL;DR: It can be undefined, but it won't be used in that case

File: src/main/webapp/i18n/de/programmingExercise.json [hunk 4]
```diff
@@ -366,6 +366,7 @@
                     "title": "Code-Style normalisieren",
                     "tooltip": "Standardisiere Zeilenumbrüche und Kodierung für Konsistenz"
                 },
+                "successMessageRepo": "Das Repository wurde erfolgreich exportiert. Die generierte zip-Datei wird jetzt heruntergeladen.",
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/i18n/de/programmingExercise.json:369]: You're right. This is consistently wrong across all translation files, though. If you'd like, I can open a separate PR that updates the translations from `zip` to `ZIP` where needed

File: src/main/java/de/tum/in/www1/artemis/service/export/ProgrammingExerciseExportService.java [hunk 5]
```diff
@@ -439,6 +439,36 @@ public Optional<File> exportStudentRequestedRepository(long exerciseId, boolean
...
+        var blankExportOptions = new RepositoryExportOptionsDTO();
+        Path outputDirectory = fileService.getTemporaryUniquePathWithoutPathCreation(repoDownloadClonePath, 5);
+        try {
+            Path zipFile = createZipForRepositoryWithParticipation(programmingExercise, participation, blankExportOptions, outputDirectory, outputDirectory);
+            if (zipFile != null) {
+                return Optional.of(zipFile.toFile());
+            }
+        }
+        catch (IOException e) {
+            String error = "Failed to export the student repository of programming exercise " + exerciseId + " and participation " + participation.getId();
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/java/de/tum/in/www1/artemis/service/export/ProgrammingExerciseExportService.java]: I updated it

File: src/main/webapp/i18n/de/programmingExercise.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/i18n/de/programmingExercise.json:369]: Alright, I'll ask him tomorrow :+1:

**Review by coolchock — APPROVED** [review]
Re-approve after minor code changes

**Review by JohannesStoehr — APPROVED** [review]
Latest code changes look good to me

**Review by JohannesWt — APPROVED** [review]
Tested on TS2 - works as expected 👍
